sae/consensus.go

```diff
@@ -32,15 +36,89 @@ func (vm *VM) SetPreference(ctx context.Context, id ids.ID, bCtx *block.Context)
 //
 // [accepted]: https://github.com/ava-labs/avalanchego/tree/master/vms#block-statuses
 func (vm *VM) AcceptBlock(ctx context.Context, b *blocks.Block) error {
-	// TODO(arr4n) implement settlement, updating of database canonical block,
-	// head, etc.
-	vm.lastAccepted.Store(b)
-	return vm.exec.Enqueue(ctx, b)
+	// Recall the terminology and ordering from the invariants document:
+	// - (D)isk then (M)emory then (I)nternal then e(X)ternal.
+	// - B accepted after all of B.Settles() settled
+
+	settles := b.Settles()
+	{
+		batch := vm.db.NewBatch()
+
+		// D(s \in S); yes, I know it's in a batch
+		if s := settles; len(s) > 0 {
+			// Note that while we treat SAFE and FINALIZED identically, there is
+			// no notion of the former in [rawdb]. Furthermore, the LAST label
+			// is reserved for the last executed so it too isn't persisted here.
+			rawdb.WriteFinalizedBlockHash(batch, s[len(s)-1].Hash())
+		}
```

Comment:
While slightly less efficient, I feel like it's more clear
```suggestion
		if len(settles) > 0 {
			// Note that while we treat SAFE and FINALIZED identically, there is
			// no notion of the former in [rawdb]. Furthermore, the LAST label
			// is reserved for the last executed so it too isn't persisted here.
			rawdb.WriteFinalizedBlockHash(batch, b.LastSettled().Hash())
		}
```
Technically, we could even remove the `if` and always write the finalized block hash, but reducing DB IO is always a good thing (even at the slight reduction in readability).

Reply:
> While slightly less efficient, I feel like it's more clear
💯